### DISCUSSION TITLE
No TLS in some devices - minimal upgrade

### DISCUSSION THREAD

**Discussion by dac2020** [arendst/Tasmota#16376] [Q&A] [answered]
Do some tasmota devices not have TLS capability?

I ordered a couple of these...
https://www.amazon.com/dp/B0828BH1ZK

After updating them with tasmota then using the upgrade function, I still see "MINIMAL firmware, please upgrade".

Also, is that why I'm not seeing any TLS option? I'm using port 8883 with SSL cert for my broker.
Does it mean I have to build my own firmware or something else?

**Comment by sfromis** [comment, marked as answer]
The standard precompiled Tasmota builds do not include the optional feature of TLS support.

What happens for your upgrade depends on details of what *exactly* you did, and network access to the compressed .gz binary (always use those) you try to upgrade to.

**Reply by barbudor** [reply]
Your browser may have got the intermediate minimal firmware
That doesn't mean that the upgrade didn't carry on
Just refresh the browser window

**Reply by dac2020** [reply]
> The standard precompiled Tasmota builds do not include the optional feature of TLS support.
> 
> What happens for your upgrade depends on details of what _exactly_ you did, and network access to 
>the compressed .gz binary (always use those) you try to upgrade to.

I simply used the Upgrade firmware option on the two relays I first purchased and both of those have TLS.
I did not provide any file, the devices used this URL;
http://ota.tasmota.com/tasmota32/release/tasmota32.bin

**Comment by barbudor** [comment]
TLS is a hungry feature which is generally not needed on a private LAN
So as ESP8266 has limited resources it is not included by default in any official binaries
As ESP32 has more capacities, it has recently been decided to enable it by default on every ESP32 binaries

So yes, for ESP8266, you will need self compile binary.
I'm using that for a site with very little number of devices connecting to a cloud based broker.

**Reply by dac2020** [reply]
I refreshed and that warning is gone now. Thank you.
However, still no TLS option available but now I see why based on your replies.

I'm not testing on private LAN anymore. I have a broker set over the Internet as I have different locations where I'd like to control things.
The two relay boards I first purchased both upgraded using the upgrade function and both have TLS capability, that's why I asked.

Darn, I don't think I have the time I need right now to take on building my own firmware as well. This has been one heck of a lot of learning as it is :).

**Reply by dac2020** [reply]
So I guess I have to also find out if devices handle TLS before buying?

**Reply by barbudor** [reply]
Not the right question. All devices support TLS, just not built in by default in Tasmota for ESP8266

The right question if you don't want self compile is does the device uses an ESP32 variant (ESP32, ESP32c3 or esp32-solo1)

**Reply by dac2020** [reply]
I'm not saying I don't want to learn to self compile, I'm saying I have a lot on my plate as it is and been trying to learn this as a sideline :).
Not sure I can take that on as well but I'll look at it if that's my only option.

The thing I don't understand is I did nothing special with the two relays I first started with. Both were upgraded directly from the ota.tasmota.com site and have TLS.

**Reply by barbudor** [reply]
Because they are based on ESP32
The plug on the contrary are based on ESP8266
The processor is listed in the information page of the device

**Reply by dac2020** [reply]
Ok, so I have to buy specifically ESP32 stuff.

**Reply by barbudor** [reply]
The processor is listed on the template page of the device

**Reply by sfromis** [reply]
The `zbbridge` build for ESP8266 also includes TLS, but not much in sensors, of course....

**Reply by dac2020** [reply]
I'll have to note that next time I buy hardware so it doesn't happen again.
I also ordered a Shelly EM and can't seem to confirm if that is ESP25.

Guess I'm stuck with these plugs I cannot really use unless I can learn to compile next.
I'll look into compiling as soon as I can.

**Reply by dac2020** [reply]
> The `zbbridge` build for ESP8266 also includes TLS, but not much in sensors, of course....

What's a zbbridge?

**Reply by sfromis** [reply]
It is one of the precompiled builds from the OTA site you already know of. It is for the Sonoff Zigbee Bridge, but also works without that specific device.

**Reply by dac2020** [reply]
Too new to fully understand that :). 
OTA is the site that has the firmware but do you mean there is a firmware I could use on this device that will have the basics plus TLS? I'll go look at that device again on the hardware page. Maybe I missed something.

https://templates.blakadder.com/avatar_AWP12L.html

**Reply by dac2020** [reply]
Ah, now I see what you mean. There's more info on the OTA site. I didn't know that before. I'll dig around for the device I have.

**Reply by dac2020** [reply]
I think you are saying to use this?
http://ota.tasmota.com/tasmota/release/tasmota-zbbridge.bin.gz
Asking before I jump and upgrade so I don't brick these things :).

If you are saying that, can you let me know how I would find out which builds are for which devices?

**Reply by s-hadinger** [reply]
Be careful because this firmware is flashed with a 2MB layout which would fail with devices with only 1MB

**Reply by sfromis** [reply]
As long as you have wired reflashing as a backup plan, no harm in trying that binary, **if** it works. Of course, preferable would be a custom build.
https://tasmota.github.io/docs/Firmware-Builds/

**Reply by dac2020** [reply]
Information shows this. Looks like a 1MB I guess.

![image](https://user-images.githubusercontent.com/9727841/187094550-dada0c94-eb9c-4bcc-862b-c6f0d2f7d561.png)

**Reply by dac2020** [reply]
I guess I'm a little discouraged today. It's pretty hard to know what hardware works with tasmota. 
Not the fault of tasmota of course but manufacturers and distributors changing things all the time.
I'm finding myself never being fully sure of what to buy or could even keep buying if it works for me.
For example, this is kinda neat. https://www.amazon.com/gp/product/B08F8PT7YR
But, will it work? No idea since there are a couple of mentions of tasmota but sellers barely ever reply so it's not clear. 

I was very excited to try and learn all this new stuff but buying the hardware seems to be the limiting factor of it all.

**Reply by barbudor** [reply]
You are discovering this world and are overwhelmed by all you have to learn
So you need to keep on focusing and go step by step just like you've done those last weeks.

The industry just doesn't care about us and what we are doing. We are too small a market for them. Some may even consider this goes against their business as they invested into infrastructure (cloud, apps) to support their product and we are offerring to free users from that.
Most of the products on the market are sacrifiable. They manufacture them, throw them to the market and just don't care. If the company gets down, nobody pays anymore for the cloud service, everything is shut down abruptly and devices becomes unusable. This happenned before in the past and will happen again. You can't expect the general mass market to have the approach we have.

A very small number of brand like Athom and others has decided to embrace that little small market, find a white label manufacturer and sell products with Tasmota pre-flashed. 

So if you don't want to try random devices from AMZ and Ali to discover 70% of them are not compatible, just stick on brands that are stable and won't go away tomorrow : Sonoff and Shelly are probably the most trustable at that time for not changing the design of the product. And the template site clearly identify which have ESP8266 or ESP32.
Forget about those random plugs, strip or plugs for the time being, you have other things to do/learn.

At least, compile your build should be your next target and you should have a look at TasmoCompiler which will allow you to build that TLS binary for ESP8266 in a matter of a minute without learning Git, C++ and VisualCode.

Just my 2c

**Reply by dac2020** [reply]
I completely understand what you're saying above and those reasons are exactly why I am excited to come across projects like this. These projects can unify hardware at the owner level rather than seller/distributor that offers yet another unknown app that you have to install. 

The problem is how frustrating it is to find hardware in a market that as you said, is unwilling to make this easy or standardize so that alternative fw can be installed because they all want to control their buyers. 

I do need plugs and things that come in their own cases as I'm only automating, not getting into hardware and developing robotics or anything like many here might be. Am trying to find devices that can be flashed without taking them apart or wiring into them. That is my goal.

Asking on Amazon is useless and looking at the US list is nice to see but most of those items no longer even found on Amazon. For me, I need a reliable source such as Amazon so I can send the stuff back if it doesn't work otherwise 
I buy from ebay and aliexpress etc and take my chances. I already sent back three devices.

I'll take a look at the compiler but not sure I have the time or patience to brick a bunch of things let alone afford it. I guess I do have these plugs I can't use as testers.

I want to learn this because I want to spread this to others and let as many as possible know that there are alternatives to using a bunch of different cloud apps when you can not only pull all these things together in a unified way but have full control without invasive profiling too.

Even if they don't want to learn, if I know how, I can do it for them and helping them to get off the profiling/app cloud would be a small win for me :).

**Reply by sfromis** [reply]
If you mainly want to automate, without having to know internal details of devices, like if it has a MCU compatible with Tasmota (this is becoming less common by the month), you could also look towards Zigbee. This concept is also reasonable for inhouse automation, where you can have a coordinator running open source software, including Tasmota. No dependence on external sites. No task of hacking a device to flash alternate firmware.

While Zigbee devices do not give you the detailed on-device control available when flashing Tasmota (hence my preference is still Wifi+Tasmota when available), you do have a lot of device options available, and mostly they are workable with open source coordinator software. Tuya devices may (as usual) be the most troublesome, taking extra work to get supported. Of course, practical details of creating automations are different....

**Reply by dac2020** [reply]
At this moment, I am mainly interested in using hardware that gets people off these apps. 
I don't see myself 'developing' or needing too many custom changes. Just being able to control hardware without apps is good.

Mainly interested in finding devices I don't have to open to connect to. That may change but my initial goal is as an end user able to help others to start using alternative home control devices.

I will install the compiler and look at that to see if it's something I can handle If that is what will allow me to use more hardware, then it's worth the effort.

>If you mainly want to automate, without having to know internal details of devices, like if it has a MCU compatible with 
>Tasmota (this is becoming less common by the month), 

What do you mean by this comment? Do you mean it's harder for tasmota to exist because manufacturers are moving to different chips?

**Reply by sfromis** [reply]
I mean that the concept of replacing firmware on commercial devices obviously gets harder when vendors move to chips not well-supported by tinkerers (basically supported by the Arduino ecosystem). While Tasmota (and other comparable replacement firmware) is great when feasible, this is focused at people who **themselves** are willing to take the trouble and develop the understanding to make things work, not so much at people who just want something to work without themselves getting deep in the details.

And for people wanting to get "cloudless" without becoming too technical, automation via Zigbee (maybe Matter if/when that starts to matter) is a reasonable strategy to consider. One drawback of the legacy ESP8266 is that it is just not good for battery-operated devices, giving some notable weaknesses in hardware availability, also outside of the already quite limiting platform replacement issue.

Tasmota can still exist with Zigbee, as you can run a coordinator based on Tasmota, including MQTT-TLS if wanted.

**Reply by dac2020** [reply]
I'm not giving up, just frustrated how hard it is to find hardware. Will keep working on it and try the compiler.

**Comment by dac2020** [comment]
I guess my question is also, does anyone know some plugs that for sure run tasmota with TLS I can get on Amazon now.
I'm looking for minimum dual plug devices where each can be controlled.

**Reply by sfromis** [reply]
For TLS, It is not a question of what plug. As long as it has ESP8266 or ESP32, the TLS support in Tasmota can be used. For ESP8266 devices, you need to build a custom binary. In both cases, you obviously need the details for a TLS connections.

**Reply by Noschvie** [reply]
Have a loot to this : [Tasmota Steckdose NOUS A1T - Kleinste Wlan ESP8266 smart plug](https://www.amazon.de/Tasmota-Steckdose-NOUS-A1T-stromverbrauch/dp/B0054PSIPA/ref=asc_df_B0054PSIPA/?tag=googshopde-21&linkCode=df0&hvadid=546520626712&hvpos=&hvnetw=g&hvrand=1262317463588225332&hvpone=&hvptwo=&hvqmt=&hvdev=c&hvdvcmdl=&hvlocint=&hvlocphy=9062729&hvtargid=pla-1604264872293&psc=1&th=1&psc=1)
The plug will be up and running in 2 minutes ...

**Reply by farrad** [reply]
@Noschvie I am actually struggling to get this one up and running.
1. Upgraded to Tasmota 12.3.1 from: http://ota.tasmota.com/tasmota/release/tasmota.bin.gz
2. The issue I am facing is `OpenSSL Error[0]: error:1402542E:SSL routines:ACCEPT_SR_CLNT_HELLO:tlsv1 alert protocol version` error on Mosquitto Broker

**Reply by barbudor** [reply]
As already explained, you upgraded to a Tasmota binary which doesn't include TLS
Read the docs, you need to compile your own binary to have the feature on esp8266

**Comment by dac2020** [comment]
I'm working on this. 
https://tasmota.github.io/docs/Create-your-own-Firmware-Build-without-IDE/#prepare-a-platformio-core-environment-contained-in-a-folder

I'm building my first one but it's just a test. I'll post a new question.

**Reply by Noschvie** [reply]
Have a look to the Online Compiler : https://tasmota.github.io/docs/Gitpod/

**Reply by barbudor** [reply]
@dac2020 the shortest path is not to install de development environment unless you are at ease and have experience with that.
But try TasmoCompiler (not just the "compiler" but TasmoCompiler which is a on-line, check-box driven configuration to build a binary without the hassle of installing anything on your computer

https://github.com/benzino77/tasmocompiler

With regards of building for ESP8266 or ESP32, you choose 'tasmota' for esp8266 and 'tasmota32' for ESP32

**Reply by dac2020** [reply]
> Have a look to the Online Compiler : https://tasmota.github.io/docs/Gitpod/

Nice, thanks. I'll give it a try.

**Reply by dac2020** [reply]
> @dac2020 the shortest path is not to install de development environment unless you are at ease and have experience with that. But try TasmoCompiler (not just the "compiler" but TasmoCompiler which is a on-line, check-box driven configuration to build a binary without the hassle of installing anything on your computer
> 
> https://github.com/benzino77/tasmocompiler

Sorry, I'm not quite understanding this. It looks like code I have to install on my own server to get a web based compiler?

**Reply by sfromis** [reply]
Certainly no. That is just one option, but you can also run it in the browser without any install, if you just read the page and do what it says.
https://github.com/benzino77/tasmocompiler#how-to-start-using-tasmocompiler
It starts with a "Super easy way" :smiley_cat:

**Reply by dac2020** [reply]
I only skimmed it since I'm doing too many things at the same time but planning to go back and read closely asap.

**Comment by barbudor** [comment]
@dac2020 Just came across a discussion about a device that could be of interrest for you.
Based on ESP32, so standard tasmota32 has TLS ;)

https://github.com/arendst/Tasmota/discussions/16394

**Reply by dac2020** [reply]
Thank you for letting me know. I'll go look. I like that I was able to add TLS to these plugs.
https://github.com/arendst/Tasmota/discussions/16384

**Reply by dac2020** [reply]
Nice units, great prices.

**Reply by dac2020** [reply]
More to learn. I used the build I did with TLS to upgrade a shelly em but there's no current information.
I found this and it seems the template and everything is good so I guess it means I need a more custom firmware for this one.
https://github.com/arendst/Tasmota/issues/14261

**Reply by barbudor** [reply]
Did you cut the power to the Shelly to hard-reset it as recommended? 
Soft restart is not enough.

**Reply by dac2020** [reply]
Yes, I read all the comments and did all steps.
I checked the template just like the post suggested as well but no metering.

My build was really simple, './compiler.sh tasmota' and I used that build. 
I'm assuming I need to rebuild that with some option or directive that is for the shelly?

**Reply by barbudor** [reply]
If you have build default `tasmota.bin.gz` with just adding TLS then everything for Shelly EM should be there
Can you please paste here the result of `backlog template ; module ; gpio ; i2cscan ; i2cdriver` ?

**Comment by dac2020** [comment]
```
backlog template
22:09:41.397 CMD: backlog template
22:09:41.429 RSL: RESULT = {"NAME":"Shelly EM","GPIO":[0,0,0,0,0,0,0,0,640,3456,608,224,0,1],"FLAG":0,"BASE":18}```

module
22:06:22.952 CMD: module
22:06:22.956 RSL: RESULT = {"Module":{"0":"Shelly EM"}}

gpio
22:07:19.257 CMD: gpio
22:07:19.261 RSL: RESULT = {"GPIO17":{"0":"None"}}

i2cscan
22:08:01.641 CMD: i2cscan
22:08:01.661 RSL: RESULT = {"I2CScan":"Device(s) found at 0x38 0x51"}

i2cdriver
22:09:19.699 CMD: i2cdriver
22:09:19.704 RSL: RESULT = {"I2CDriver":"7"}
```

**Reply by barbudor** [reply]
Looks OK.
No idea for now

**Comment by dac2020** [comment]
I doubt it's a bad device since it was working fine just before changing it to tasmota. 
In fact it is still working, just no power monitoring.
![image](https://user-images.githubusercontent.com/9727841/190516917-80df87b8-beac-4812-8e56-d7ed2ffdb6bb.png)

**Comment by dac2020** [comment]
If this helps...

I used the Shelly EM build 
https://github.com/yaourdt/mgos-to-tasmota

Then I upgraded the firmware with my own build since it has TLS and the above one didn't and is minimal.

**Reply by barbudor** [reply]
I'm not sure how you could have used mgos-to-Tasmota from minimal as this is a procedure intended to OTA from Shelly firmware to Tasmota without using serial.
If you have serial-flashed tasmota you don't need mgos-to-tasmota

May be retry a serial reflash of official Tasmota with full erase.
Configure the device without mqtt to get proper power values on web page.
Once this is ok, upgrade to your firmware and check that power values are still here.
Move on configuring mqtt only if those steps are ok

**Comment by dac2020** [comment]
I found a doc online that showed how to use OTA to change it from shelly to tasmota. 
It showed to do this;
http://192.168.254.204/ota?url=http://dl.dasker.eu/firmware/mg2tasmota-ShellyEM.zip
This worked, it changed the firmware to tasmota.

I looked and don't have anything small enough to fit those pins. 

Is there anything I can do from the GUI for now?

I've disabled MQTT for now as you said so should I use?
Using the default Firmware upgrade from tasmota shows no power values.
If using my own build, again, without MQTT enabled, shows no power values.

```
Program Version | 12.1.1(tasmota)
Build date /time 2022-08-25T11:33:55
Core/SDK version 2_7_4_9/2.2.2-dev(38a443e)
Flash write count 48 at 0xFA000
Boot count 19

ESP Chip Id | 16572811 (ESP8266EX)
Flash chip ID 0x1620C2
Flash size 4096 KB
Program flash size 1024 KB
Program size 624 KB
Free program space 376 KB
Free memory 25.7 KB
```

**Reply by barbudor** [reply]
So if you can't serial flash, may be you can try
- use official tasmota.bin.gz
- command : `reset 2` will erase all configs to insure restarting from scratch, including wifi
- configure minimum : wifi, template (no mqtt)
- power cycle de device (disconnect from mains, wait 30 sec, reconnect)

=> what does it do ?

**Comment by dac2020** [comment]
Sure, I can try that. I'll look on Amazon and see if I can order a cable for those devices too since I'll likely use them now.
The official, that's the one in the Firmware update right? I can use that URL to download the official latest.

**Reply by barbudor** [reply]
Yes, from http://ota.tasmota.com

**Reply by dac2020** [reply]
I followed the steps exactly and still no power readings.

**Reply by barbudor** [reply]
No more idea
May be open a new discussion ? to attract people who may own a Shelly EM

**Reply by dac2020** [reply]
Sure, I will, thanks.
